infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/EphemeralWorkspaceAdapter.java

```diff
@@ -0,0 +1,129 @@
+   * @throws InternalInfrastructureException
+   */
+  public boolean isEphemeral(String workspaceId) throws InternalInfrastructureException {
+    try {
+      WorkspaceImpl workspace = workspaceManager.getWorkspace(workspaceId);
+      String mountSources = workspace.getConfig().getAttributes().get(MOUNT_SOURCES_ATTRIBUTE);
+      return "false".equals(mountSources);
+    } catch (NotFoundException | ServerException e) {
+      throw new InternalInfrastructureException(
+          "Failed to load workspace info" + e.getMessage(), e);
+    }
+  }
+
+  public void provision(KubernetesEnvironment k8sEnv, RuntimeIdentity identity)
+      throws InfrastructureException {
+    for (Pod pod : k8sEnv.getPods().values()) {
+      PodSpec podSpec = pod.getSpec();
+      for (Container container : podSpec.getContainers()) {
+        String machineName = Names.machineName(pod, container);
+        InternalMachineConfig machineConfig = k8sEnv.getMachines().get(machineName);
+        Map<String, Volume> volumes = k8sEnv.getMachines().get(machineName).getVolumes();
+        addMachineVolumes(identity.getWorkspaceId(), pod, container, volumes);
+      }
+    }
+  }
+
+  private void addMachineVolumes(
+      String workspaceId, Pod pod, Container container, Map<String, Volume> volumes)
+      throws InfrastructureException {
+    if (volumes.isEmpty()) {
+      return;
+    }
+
+    for (Entry<String, Volume> volumeEntry : volumes.entrySet()) {
+      final String volumePath = volumeEntry.getValue().getPath();
+      final String volumeName =
+          LOGS_VOLUME_NAME.equals(volumeEntry.getKey())
+              ? volumeEntry.getKey() + '-' + pod.getMetadata().getName()
+              : volumeEntry.getKey();
+
+      final String uniqueVolumeMountName = Names.generateName("ephemeral-");
+
+      // binds pvc to pod and container
+      container
+          .getVolumeMounts()
+          .add(
+              newVolumeMount(
+                  uniqueVolumeMountName,
+                  volumePath,
+                  getSubPath(workspaceId, volumeName, Names.machineName(pod, container))));
+      addEmptyDirVolumeIfAbsent(pod.getSpec(), uniqueVolumeMountName);
+    }
+  }
+
+  private void addEmptyDirVolumeIfAbsent(PodSpec podSpec, String uniqueVolumeMountName) {
+    if (podSpec
+        .getVolumes()
+        .stream()
+        .noneMatch(volume -> volume.getName().equals(uniqueVolumeMountName))) {
+      podSpec
+          .getVolumes()
+          .add(
+              new VolumeBuilder()
+                  .withName(uniqueVolumeMountName)
+                  .withNewEmptyDir()
+                  .withMedium("Memory")
+                  .endEmptyDir()
+                  .build());
+    }
+  }
+
+  private String getSubPath(String workspaceId, String volumeName, String machineName) {
+    // logs must be located inside the folder related to the machine because few machines can
```

Comment:
It seems that a lot of code is inspired by strategies code. Would be nice to reuse the code if possible.

Reply:
you are right, but reusing strategies code might be not really straight-forward

Reply:
Then let's go with a simpler approach. Time will show whether we would need to refactor that later